### DISCUSSION TITLE
Note functions tests, implementation and such

### DISCUSSION THREAD

**Discussion by ikusteu** [librocco/librocco#91] [Ideas]
@fadwamahmoud is already working on this, but I'm creating this discussion so that we can voice opinions/suggestions more easily.

## The problem

We should create functions to communicate with the DB (in our case, we would be using couch DB and pouch DB).

@silviot suggested to start by writing the simplest (testable) functionality as a first pass and have them simply write to a variable (no db). Afterwards, we would leverage the tests and expand on functionality as well as include interface to communicate with the DB.

### The idea

My suggestion is the following:

- create functions in charge of certain app operations: createNote, updateNote, addBooks, removeBooks, deleteNote, commitNote, etc. _(the list of functions is hypothetical and can be shrunk/extended as needed, as well as defined in more detail)_
- the aforementioned functions should not concern themselves with what db they're writing to, or how the actual communication with the DB should be handled, that behaviour should be handled by a DB interface 
- the implementation of an interface might vary depending on which db we're using to communicate with, but a standard should be implemented

With respect to the suggestions above, it would make sense for the functions _(createNote, addBooks, etc.)_ to look something like this:
```typescript
const createNote = async (db: DBInterface) => {
  const noteId = await db.create("notes")
  return noteId
}
const getNote = async (db: DBInterface, noteId: string) => {
  const note = await db.read(`note/${noteId}`)
  return note
}
```
Now the actual communication with the db will vary depending on the implementation of the `DBInterface`, for example:
```typescript
// The simplified interface, detailed definition is TBD 
interface DBInterface {
  // 'create' method accepts a collection and returns a string id of the newly created document 
  create: (collection: string) => Promise<string>
  // 'read' method accepts a document id and returns found document entry or `null` if not found
  read: (documentPath: string) => Promise<Document | null>
}

// in-memory (variable based implementation)
const record = {}
const inMemoryDB: DBInterface = {
  create: async (collection) => {
    const id = uuid()
    record[collection] = {
      ...record[collection],
      [id]: { /* an empty document */ }
    }
  }
  read: async (documentPath) => {
    // Rewrite "/" separated path to JSON-friendly: "." separated 
    const path = documentPath.split("/").join(".")
    return record[path] || null
  }
}

// couch db implementation
const couchDB: DBInterface = {
  create: async (collection) => {
    const id = uuid() // or maybe let the db generate an id, this is only an example anyway
    /* add a new document to couch db using its API */
  }
  read: async (documentPath) => {
    /* get the document from couch db using its API */
  }
}
```

**_Please note that the example above ☝️ is just that...an example and the clear definition of the interface as well as its methods and app note functions should be defined in this discussion 👇  and throughout the development iterations._**

### Interface definitions (suggestion)

My suggestion for the interface is to start with a simple CRUD and expand later when necessary (we'll definitely need to expand to accommodate store subscriptions, but that should be done in one of the iterations to follow):
```typescript
interface DBInterface { // I'm not hell bent on `DBInterface` naming either, but I think it makes sense
  // Create method resolves to an 'id' string of the new document
  // I feel that this should be handled by the DB implementation, rather than the caller
  create: () => Promise<string>
  // Read reads from the DB and resolves to found document, or 'null'
  // 'documentId' can easily be documentPath (string), or even path fragments: `(pathFragments: ...string) => {...}`
  // but should serve as a way to find the document in the db and, once defined, be standardised across implementations
  read: <T extends any>(documentId: string) => Promise<T | null>
  // Update updates a given entry with the data passed in
  // the 'documentId' part is the same as mentioned above
  // { merge: boolean } should be accepted to differentiate between overwriting a document and updating only the provided fields
  // Update resolves to an updated document, or throws a "document not found" error
  update: <T extends any>(documentId: string, update: Partial<T>, options: { merge: boolean }) => Promise<T>
  // Delete removes a document from the DB and resolves to deleted document id
  // Throws if document not found
  delete: (documentId: string) => Promise<string>
}
```

As for the note functions I suggest for them to be written in this way: 
```typescript
const createNote = (db: DBInterface) => async () => {
  // Note creation functionality
}
const addBooks = (db: DBInterface) => async (noteId: string, booksToAdd: BookStock[]) => {
  // Leverage 'db.update' to add specific fields, or increase count of a book stock
}
```
What's important (in my suggestion) is for the note function to satisfy the following (NoteOperationHOF) interface:
```typescript
interface NoteOperation<P extends any[], R extends any> {
  (...P): Promise<R>
}
interface NoteOperationHOF<P extends any[], R extends any> {
  (db: DBInterface): () => NoteOperation<P, R>
}
```
This way, every note function would be a higher order function accepting a db interface and returning an (async) note operation function enclosed with the `db` interface to use for communication with the database:
```typescript
// 'never[]' is an empty array or, in this case, no parameters for a returned function
const createNote: NoteOperationHOF<never[], string> = (db) => async () => {
  // Note creation functionality
}
```

### Testing

As this logic is implemented, the testing becomes trivial and highly reusable:
1. We start by writing a simple `DBInterface` that writes to a variable (kinda like the one described above)
2. We write tests for all required note functions, passing the interface from the first point to the functions (as `db`) and testing the resulting state using the same `db` interface (using `db.get`), example:
```typescript
const db = /* a variable oriented db interface */
test("'addDocument' should add document to the database using the provided db interface", async () => {
  const document = {foo: "bar"}
  await addDocument(db)(docId: "test-document", document)
  const docInDB = await db.get("test-document")
  expect(docInDB).toEqual(document)
})
```
3. Make the tests from 2nd point into a table test run:
```typescript
const runTableTests = (interfaces: {name: string, db: DBInterface}[]) => interfaces.forEach(({title, db}) => {
  describe(`Test "${name}" as db interface`, () => {
    test(/* Test from example before */)
    /* Additional function tests */
  })
})
runTableTests([
  {name: "In memory DB", db: inMemoryDB},
  {name: "Couch DB", db: couchDB},
])
```
4. Each time we add a new db interface (for CouchDB, PouchDB, SQL, Firestore, etc.) we just need to add additional table test case using that particular implementation and the run the same tests over the new interface, with no test code itself changing:
```typescript
// Following the example above, say we want to add "pouchDB interface", our test run would look something like this
runTableTests([
  {name: "In memory DB", db: inMemoryDB},
  {name: "Couch DB", db: couchDB},
  {name: "Pouch DB", db: pouchDB},
])

```
5. Only when we update `DBInterface` definition (by expanding to allow subscriptions for instance) do we need to change the test code

### Tasks (things to define)

- [ ] Define exactly how each of the CRUD methods should behave (what should they accept as param, what should they resolve to, etc.)
- [ ] Define note actions (mentioned in the previous discussion)
- [ ] Implement note actions and test them (TDD) against an in-variable db interface
- [ ] Wrap tests into a table test runner (somewhat as mentioned above)

### Enhancements (next steps)

- add db interface implementations for couchDB and pouchDB (from what I understand, it could be the same, only initialised with different params)
- extend interface to enable store subscriptions (to live updates): I suggest an additional method returning an observable or observable-like (or plain and simple: a svelte store)

**Comment by silviot** [comment]
I agree in general with the whole structure of this proposal.
I believe though the focus, in this step, should be in writing a minimal useful test.
Also, I think the `DBInterface` should allow one to create a new database. Each test should start with an empty db.
A starting point for testing could be:
```
Create a DB
List all notes in the DB
Make sure there are none
Create a note with name "FooBar"
List all notes in the DB
Make sure there is one, only one, and that its name is "FooBar"
Remove the note
List all notes in the DB
Make sure there are none
```
and
```
Create a DB
Create a note named "TestNote"
List all books in "TestNote"
Make sure there are none
Add a book with ISBN "12345"
List all books in "TestNote"
Make sure there is only one, with ISBN "12345" and quantity 1
Add two copies of a book with ISBN "12345"
List all books in "TestNote"
Make sure there is only one, with ISBN "12345" and quantity 3
Remove one copy of a book with ISBN "12345"
List all books in "TestNote"
Make sure there is only one, with ISBN "12345" and quantity 2
Add a book with ISBN "6789"
List all books in "TestNote"
Make sure there are two, ISBN "12345", quantity 2 and ISBN "6789", quantity 1
```
